### PR TITLE
Fix KMS secret rotation follow-up cases

### PR DESCRIPTION
**Pull request by Oluwatobi-Mustapha**
Refs #1533

Follow-up to #1676 addressing the remaining review comments on the secret/key rotation planner.\n\nChanges:\n- Classify equivalence-backed secret rotation plans as `kms_related` when the matched Secrets Manager metadata only has `KMSKeyID` or an alias and no resolved `KMSKeyARN`.\n- Canonicalize KMS refs through the KMS reachability index before checking whether multiple secrets share the same key, so ARN/alias/key-id forms are grouped together.\n- Add regression coverage for alias-only KMS classification and shared-key readiness gates across mixed KMS ref forms.\n\nValidation:\n- `go test ./internal/api -run "TestAWSSecretKeyRotationPlanFromMetadata(PreservesKMSKeyID|CanonicalizesSharedKMSRefs)|TestAWSSecretKeyRotationTypeKeepsAWSNativeSecretFindings"`\n- `go test ./internal/api -run "TestAWSSecretKeyRotation|TestGetAWSSecretKeyRotation|TestRouterAWSSecretKeyRotation"`

---
## Summary by cubic
Fixes KMS secret rotation edge cases by canonicalizing, scoping, and deduplicating KMS references. Avoids cross‑account alias leaks, picks the right key per account/region, and keeps KMS vs provider‑key classification accurate.

- **Bug Fixes**
  - Canonicalize KMS refs via the reachability index. Scope alias lookups and target key resolution by account/region, avoid cross‑scope fallback when a bare alias can’t be resolved, and drop ambiguous unscoped refs.
  - Preserve types: keep `provider_key` findings as `provider_key` even with KMS metadata; classify alias‑only or `KMSKeyID`‑only secrets and external‑provider KMS equivalence findings as `kms_related`.
  - Deduplicate canonical KMS buckets and per‑secret entries so ARN/alias/key‑id forms aren’t double‑counted, and only add the shared‑key gate when multiple distinct secrets share the same canonical key.

<sup>Written for commit 64319dacfb368dacd54c4728554c7dcc434cd8d8. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/identrail/identrail/pull/1677?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: internal/api/aws_secret_key_rotation.go [hunk 1]
```diff
@@ -704,6 +706,34 @@ func awsSecretKeyRotationKMSRef(secret AWSSecretsManagerMetadataRecord) string {
...
+	if len(secretsByKMS) == 0 {
+		return secretsByKMS
+	}
+	out := map[string][]AWSSecretsManagerMetadataRecord{}
+	for ref, secrets := range secretsByKMS {
+		canonical := awsSecretKeyRotationCanonicalKMSRef(ref, kmsByARN)
+		if canonical == "" {
+			continue
+		}
+		out[canonical] = append(out[canonical], secrets...)
```
[the remark this answers is not in the document]
Comment (Oluwatobi-Mustapha) [reply internal/api/aws_secret_key_rotation.go]: Fixed in f4af0c84. `awsSecretKeyRotationCanonicalSecretsByKMS` now deduplicates secrets within each canonical KMS bucket using the secret ARN/node/name before shared-key readiness checks run. Added `TestAWSSecretKeyRotationPlanFromMetadataDoesNotDuplicateCanonicalKMSRefs` to cover a single secret with both KMSKeyARN and KMSKeyID.

File: internal/api/aws_secret_key_rotation.go [hunk 2]
```diff
@@ -484,8 +486,8 @@ func finalizeAWSSecretKeyRotationPlan(plan AWSSecretKeyRotationPlan) AWSSecretKe
 
 func awsSecretKeyRotationType(finding AWSSecretPermissionEquivalenceFinding, secret AWSSecretsManagerMetadataRecord) string {
 	normalized := normalizeAWSRuntimeEventFilterToken(finding.EquivalenceType)
-	secretRef := strings.ToLower(strings.Join([]string{finding.SecretNodeID, finding.SecretARN, finding.SecretLabel, secret.KMSKeyARN, strings.Join(finding.SourceSignals, " ")}, " "))
-	if strings.Contains(normalized, "kms") || strings.Contains(secretRef, "kms") {
+	secretRef := strings.ToLower(strings.Join([]string{finding.SecretNodeID, finding.SecretARN, finding.SecretLabel, secret.KMSKeyARN, secret.KMSKeyID, strings.Join(finding.SourceSignals, " ")}, " "))
+	if strings.Contains(normalized, "kms") || strings.Contains(secretRef, "kms") || strings.TrimSpace(secret.KMSKeyARN) != "" || strings.TrimSpace(secret.KMSKeyID) != "" {
```
[the remark this answers is not in the document]
Comment (Oluwatobi-Mustapha) [reply internal/api/aws_secret_key_rotation.go]: Fixed in 7f72ac7f. Provider-key findings are classified before KMS metadata is considered, so workload/agent provider-key equivalence stays under `provider_key` even when the matched secret has KMSKeyID/KMSKeyARN metadata. Added regression coverage for provider-key findings with KMS-backed secret metadata.

File: internal/api/aws_secret_key_rotation.go [hunk 3]
```diff
@@ -704,6 +706,51 @@ func awsSecretKeyRotationKMSRef(secret AWSSecretsManagerMetadataRecord) string {
...
+	return strings.ToLower(strings.TrimSpace(firstNonEmptyAWSValue(secret.SecretARN, secret.FromNodeID, secret.SecretName, secret.EvidenceRef)))
+}
+
+func awsSecretKeyRotationCanonicalKMSRef(ref string, kmsByARN map[string]AWSKMSDecryptReachabilityRecord) string {
+	ref = strings.ToLower(strings.TrimSpace(ref))
+	if ref == "" {
+		return ""
+	}
+	record := kmsByARN[ref]
+	canonical := strings.ToLower(strings.TrimSpace(firstNonEmptyAWSValue(record.KeyARN, record.KeyID, firstString(record.Aliases), ref)))
```
[the remark this answers is not in the document]
Comment (Oluwatobi-Mustapha) [reply internal/api/aws_secret_key_rotation.go]: Fixed in 7f72ac7f. KMS canonicalization now normalizes alias ARNs like `arn:...:alias/name` to `alias/name` before looking up the KMS record, so alias ARN refs group with key ARN/key ID refs for shared-key readiness gates. Added regression coverage in the canonical shared KMS ref test.

File: internal/api/aws_secret_key_rotation.go [hunk 4]
```diff
@@ -704,6 +706,54 @@ func awsSecretKeyRotationKMSRef(secret AWSSecretsManagerMetadataRecord) string {
...
+
+func awsSecretKeyRotationCanonicalKMSRef(ref string, kmsByARN map[string]AWSKMSDecryptReachabilityRecord) string {
+	ref = strings.ToLower(strings.TrimSpace(ref))
+	if ref == "" {
+		return ""
+	}
+	if strings.Contains(ref, ":alias/") {
+		ref = ref[strings.LastIndex(ref, ":alias/")+1:]
+	}
+	record := kmsByARN[ref]
```
[the remark this answers is not in the document]
Comment (Oluwatobi-Mustapha) [reply internal/api/aws_secret_key_rotation.go:793]: Fixed in 5df73c35. KMS ref canonicalization now indexes scoped `account|region|ref` keys, treats unscoped aliases/key IDs as usable only when unambiguous, and resolves each secret through its own account/region context before grouping shared KMS buckets. Added `TestAWSSecretKeyRotationPlanFromMetadataScopesBareKMSAliases` to cover identical aliases reused across different accounts.

File: internal/api/aws_secret_key_rotation.go [hunk 5]
```diff
@@ -692,8 +694,20 @@ func awsSecretKeyRotationScoreForSecret(secret AWSSecretsManagerMetadataRecord)
 
 func awsSecretKeyRotationKMSIndex(kms AWSKMSDecryptReachabilityInventoryResult) map[string]AWSKMSDecryptReachabilityRecord {
 	out := map[string]AWSKMSDecryptReachabilityRecord{}
+	ambiguous := map[string]struct{}{}
 	for _, record := range kms.Records {
 		for _, ref := range awsSecretKeyRotationKMSRecordRefs(record) {
+			if scoped := awsSecretKeyRotationScopedKMSRef(record.AccountID, record.Region, ref); scoped != "" {
+				out[scoped] = record
```
[the remark this answers is not in the document]
Comment (Oluwatobi-Mustapha) [reply internal/api/aws_secret_key_rotation.go:704]: Fixed in 7e21cf68. `awsSecretKeyRotationTargetsForKMS` now uses the same secret-scoped KMS record lookup as canonicalization, so ambiguous bare aliases still resolve to the correct account/region record for target metadata. The scoped alias regression now asserts the target keeps the correct KMS node, ARN, and evidence ref.

File: internal/api/aws_secret_key_rotation.go [hunk 6]
```diff
@@ -484,13 +486,13 @@ func finalizeAWSSecretKeyRotationPlan(plan AWSSecretKeyRotationPlan) AWSSecretKe
 
 func awsSecretKeyRotationType(finding AWSSecretPermissionEquivalenceFinding, secret AWSSecretsManagerMetadataRecord) string {
 	normalized := normalizeAWSRuntimeEventFilterToken(finding.EquivalenceType)
-	secretRef := strings.ToLower(strings.Join([]string{finding.SecretNodeID, finding.SecretARN, finding.SecretLabel, secret.KMSKeyARN, strings.Join(finding.SourceSignals, " ")}, " "))
-	if strings.Contains(normalized, "kms") || strings.Contains(secretRef, "kms") {
-		return "kms_related"
-	}
 	if awsSecretKeyRotationIsProviderKeyFinding(normalized, finding.Provider) {
 		return "provider_key"
```
[the remark this answers is not in the document]
Comment (Oluwatobi-Mustapha) [reply internal/api/aws_secret_key_rotation.go:493]: Fixed in 64319dac. KMS equivalence types now classify as `kms_related` before provider-key detection, so external-provider secrets reached via `kms_decrypt_secret_equivalence` / `kms_live_grant_secret_equivalence` keep the KMS grant scoping flow. Added regression coverage for an external provider KMS finding.

**Review by unicorn-iam — APPROVED** [review]
